Add rake script to export/import miq alerts and alert profiles

branic opened the pull request:
These rake scripts provide functionality for exporting/importing several of the ManageIQ object types:

- alerts
- buttons
- customization templates
- automate domains
- policies
- provision dialogs
- reports
- roles
- service catalogs
- service dialogs
- tags
- widgets

There are a couple of supporting ruby files that provide supporting classes to parse options and ensure there are no illegal characters in file and/or directory names.

These scripts are from the [CFME RH Consulting Scripts](https://github.com/rhtconsulting/cfme-rhconsulting-scripts) and are used by Red Hat consultants to enable storing customizations in Git and maintaining customizations between environments (e.g. dev/qa/prod) for an SDLC lifecycle.

[PR 112](https://github.com/ManageIQ/manageiq-appliance/pull/112) in manageiq-appliance is dependant on this PR and provides a bash script wrapper for these rake scripts.

Fryguy commented:
@branic First off, thanks for all of this!  Just wanted to state that so you know that my comments below don't reflect that we don't appreciate all of this work 😄 

- Is the goal to just get these in as is, and then clean them up afterwards?  Or do we want to clean them up in the course of this PR?  For one, the Red Hat references and namings have to be removed. (e.g. rhconsulting)
- Rubocop and style issues will probably need to be cleaned up for much stuff.  If the goal is to just get the original in and iterate on that, we can let that go.  Also, the lib/tasks directory is probably not the right place for most of the classes and utility modules, so those would probably have to move.
- Why are there commented lines of code?  Are those optional or just unused?
- I see authorship at the top of these files, but we generally don't have authorship in files as per our license the authors are the "ManageIQ Authors".  Is it ok, if we move that?  Was there a license on the source of these that we need to adhere to?  If the author wants attribution, perhaps they should make the PR?

branic commented:
Thanks for the comments @Fryguy. This first commit was just a straight copy of the files as they exist in the rhconsulting github. I'll work on cleaning up based on your comments and the Rubocop issues.

The commented code is from keeping old code when the code changed and I can remove them as part of the cleanup effort.

Are additional commits to this PR ok for the cleanups?

Fryguy commented:
I think comments will come as the code progresses.  I think the best approach to get this in faster and easier might be to slim this down to, say, only one of the import scripts + the accompanying helper modules.  A 1700+ line PR is almost guaranteed to take a long time.  However, if you trim it down to just one script, then we can iterate the review process on just that script, uncovering all of the various comments, and get it in quickly.  Once it's in, now you have a foundation for more rake tasks (probably one per PR) and since all of the comments have been already iterated on and discussed, getting these new PRs in will be a LOT faster.

Fryguy commented:
> Are additional commits to this PR ok for the cleanups?

Yes, that is preferable

branic commented:
@Fryguy I've removed all but the script to export/import alerts and the two helper modules from this PR. There has been some re-factoring to get rid of references to rhconsulting and general cleanup. I'm not sure what the best way to fix the last of the rubocop offenses.

Hopefully this makes it easier to review and can be used as a model for getting the other rake scripts added (in seperate PR's).

All feedback is appreciated.

carbonin commented on lib/tasks/evm_import_export_illegal_chars.rb:
```diff
@@ -0,0 +1,30 @@
+  @@regex_remove_spaces = %r{[/| ]}
+
+  # Replacement character
+  @@replacement         = '_'
```
These class variables should be constants.

carbonin commented on lib/tasks/evm_import_export_options.rb:
```diff
@@ -0,0 +1,100 @@
+# @class EvmExportImportOptions
+# @brief Simple class to parse options passed in to a rake task.
+class EvmImportExportOptions
```
This class can probably be replaced by Trollop using the rake task options like https://github.com/ManageIQ/manageiq/blob/02c1f12c0f13efeb9a3faaa8c5031a3fd5a20410/lib/tasks/evm_dba.rake#L73

carbonin commented on lib/tasks/evm_import_export_alerts.rake:
```diff
@@ -0,0 +1,114 @@
+require_relative 'evm_import_export_illegal_chars'
+require_relative 'evm_import_export_options'
+
+class EvmImportExportAlerts
```
These classes should probably live in [lib/task_helpers](https://github.com/ManageIQ/manageiq/tree/02c1f12c0f13efeb9a3faaa8c5031a3fd5a20410/lib/task_helpers).

branic commented:
@Fryguy, @carbonin,

I've refactored this PR to break out the classes and functions. I also implemented Trollop for processing the arguments and removed positional arguments (for example the destination directory for an export has to be specified as an option.

carbonin commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,41 @@
+    # @param [String] str string to replace
+    # @param [Hash] options If options[:keep_spaces] == true, then keep spaces in
+    # \p str , otherwise replace spaces. (default = false, replace spaces)
+    def self.replace_chars(str, options = {})
```
I feel like if this had a better name we wouldn't need the comment.

Maybe something like `safe_filename`?

Also it would document itself better if we passed the option specifically rather than passing the whole options hash for one value. Then the method would look like `def self.safe_filename(filename, keep_spaces = false)` which is a lot more clear.

branic replied:
Agreed. I'll change the method name and definition.

branic commented on lib/task_helpers/imports/alerts.rb:
```diff
@@ -0,0 +1,29 @@
+module TaskHelpers
+  class Imports
+    class Alerts
+      class ParsedNonDialogYamlError < StandardError; end
```
This class is defined in all of the rhconsulting rake scripts, but only used in one of them. I'll remove this class as it is not used.

branic commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,41 @@
+    # @param [Hash] options If options[:keep_spaces] == true, then keep spaces in
+    # \p str , otherwise replace spaces. (default = false, replace spaces)
+    def self.replace_chars(str, options = {})
+      if options.key?(:keep_spaces) && options[:keep_spaces] == true
```
Changed this as well.

branic commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,41 @@
+      Trollop.die :directory, 'Destination directory must already exist' unless File.directory?(options[:directory])
+      Trollop.die :directory, 'Destination directory must be writeable' unless File.writable?(options[:directory])
+
+      options.delete_if { |_k, v| v.nil? || v == false }
```
I don't think it is necessary either. It came as part of the copying code from https://github.com/ManageIQ/manageiq/blob/22efb0914107963b95e064c9d7adb2aa6d905e2a/lib/tasks/evm_dba.rake for Trollop.

branic commented on lib/task_helpers/exports/alert_sets.rb:
```diff
@@ -0,0 +1,22 @@
+
+      private
+
+      def export_alert_sets(options)
```
Agreed. I'll change it.

branic commented on lib/task_helpers/imports/alert_sets.rb:
```diff
@@ -0,0 +1,29 @@
+            alertsets = YAML.load_file(fname)
+            import_alert_sets(alertsets)
+          end
+        end
```
Makes sense. I'll change to the simplified version.

branic commented:
@carbonin I've made the changes you requested. I've also adjusted the error handling and added some error handling when performing the imports.

carbonin commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,46 @@
+      options
+    end
+
+    def self.validate_directory(directory)
```
I think all of this should be done right after parsing the options.

Also this might work better if it just returned the string then you could do something like:

```ruby
error = validate_directory(options[:directory])
Trollop.die :directory, error if error
```

Then you don't have to write to standard out and you also don't need to define the exceptions.

branic replied:
I've changed this as suggested.

carbonin commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,46 @@
+    end
+
+    def self.validate_directory(directory)
+      unless directory
```
I don't think this check is necessary because directory is a required argument.

branic commented:
@carbonin @Fryguy 

Are there other changes you would like to see with this PR?

The travis-cli failure was caused by a larger issue with travis when I pushed the last commit. I'm not sure how to clear the failure.

branic commented:
@carbonin 

I've added tests for the safe_filename method, but I could use some guidance on how to handle the tests for the validate_* methods if you have any ideas or can point me to some sample code.

Also, I had to rebase my branch to be able to get a working environment which looks like it messed up the commit logs and such. I'm not sure how to fix this or what I need to do.

As always any help and feedback is appreciated.

carbonin commented on spec/lib/task_helpers/exports_spec.rb:
```diff
@@ -0,0 +1,44 @@
+describe '#safe_filename' do
```
The top level block here should be:
```ruby
describe TaskHelpers::Exports do
  ...
end
```

branic replied:
Ok. Changed.

carbonin commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,35 @@
+module TaskHelpers
+  class Exports
+    def self.new_filename(filename, keep_spaces = false)
+      new_filename = filename
```
It looks like you changed the method name and the variable name here @branic

carbonin replied:
Also, this particular line isn't needed because `String#gsub` returns a copy.

branic replied:
Oops I did change both trying to be too efficient with search and replace.

branic replied:
I had that line because the gsub returned nil if keep_spaces was true. I've changed the code to eliminate this assignment by using a ternary operator.

carbonin commented on lib/task_helpers/exports.rb:
```diff
@@ -0,0 +1,35 @@
+      new_filename = filename
+      new_filename = new_filename.gsub(%r{[ ]}, '_') unless keep_spaces
+      new_filename = new_filename.gsub(%r{[/]}, 'slash')
+      new_filename.gsub(%r{[|]}, 'pipe')
```
I think you should be able to combine these by using a hash as the second parameter (see https://ruby-doc.org/core-2.3.3/String.html#method-i-gsub)

branic replied:
I missed that when I read the doc page the first time. I've changed to use a hash.

carbonin commented on lib/task_helpers/exports/alert_sets.rb:
```diff
@@ -0,0 +1,15 @@
+        export_dir = options[:directory]
+
+        MiqAlertSet.order(:id).all.each do |a|
+          # Replace characters in the description that are not allowed in filenames
```
I don't think we need this comment

branic replied:
I've removed the comment.

carbonin commented on lib/tasks/evm_export_import.rake:
```diff
@@ -0,0 +1,37 @@
+    task :alertprofiles => :environment do
+      options = TaskHelpers::Imports.parse_options
+      TaskHelpers::Imports::AlertSets.new.import(options)
+    end
```
Does this task need the `exit` also?

branic replied:
You are correct. I've added the missing line.

carbonin commented on spec/lib/task_helpers/exports_spec.rb:
```diff
@@ -0,0 +1,44 @@
+describe TaskHelpers::Exports do
+  context 'with filename' do
```
This should probably be a `describe ".safe_filename" do`

The convention in rspec is to `describe` the class/module under test then `describe` each method you are testing then use `it ".." do` blocks for each test case on a particular method.

Then `context` can be used for shared setup between `it` or `describe` blocks.

branic replied:
That makes sense. I'll add another describe block for the method.

branic commented:
I'd like to be able to have tests for the import/export methods (and maybe even the rake tasks), but I'm not sure how to setup the tests so I need to do some more research in how to create a file to import and write and test the export within RSpec.

branic commented:
I've thought about using FakeFS [1] to be able to temporarily create the files for the RSpec tests, but the gem isn't included in ManageIQ and I don't know if a gem should be brought in just for running tests.

[1] https://github.com/fakefs/fakefs

Fryguy commented:
I'm ok with a gem that's just for tests.  Just be sure to put it into the test group in the Gemfile.

Fryguy commented:
@carbonin Once you are comfortable with the specs, you can merge.

carbonin commented:
I'm not sure you would need FakeFS here, you could probably use [Tempfile](https://ruby-doc.org/stdlib-2.3.2/libdoc/tempfile/rdoc/Tempfile.html) for import tests and [Dir.tmpdir](http://ruby-doc.org/stdlib-2.3.2/libdoc/tmpdir/rdoc/Dir.html#method-c-tmpdir) for the export tests.

If you would rather have static files for the import tests then you could also create a "data" directory in with the specs like we do for the column re-ordering tool [here](https://github.com/ManageIQ/manageiq/tree/master/spec/tools/column_ordering)

carbonin commented on spec/lib/task_helpers/exports/alerts_spec.rb:
```diff
@@ -0,0 +1,21 @@
+  let(:alert) { 'Alert_Import_Test.yaml' }
+
+  before(:each) do
+    @export_dir = Dir.mktmpdir('miq_exp_dir')
```
I think this would be better as a `let`

carbonin commented on spec/lib/task_helpers/imports/alert_sets_spec.rb:
```diff
@@ -0,0 +1,26 @@
+    options = { :source => "#{data_dir}/#{alert}" }
+    expect do
+      TaskHelpers::Imports::AlertSets.new.import(options)
+    end.to_not output.to_stderr
```
This should assert on the contents of the database rather than just whether it fails or not.

carbonin replied:
I think this applies in a few places, so I won't write it all over the place.

carbonin commented on spec/lib/task_helpers/imports_spec.rb:
```diff
@@ -0,0 +1,41 @@
+      expect(TaskHelpers::Imports.validate_source(@import_dir2)).to eq('Import source must be a filename or directory')
+    end
+
+    # rubocop:disable Style/NumericLiteralPrefix
```
I would rather just ignore the rubocop warnings than have these here.

branic replied:
I've removed the statements to disable the rubocop checks

carbonin commented on spec/lib/tasks/evm_export_import_spec.rb:
```diff
@@ -0,0 +1,29 @@
+require 'rake'
+
+describe 'evm_export_import' do
```
This file really isn't needed. Testing the prerequisites of a rake task is a bit extreme.

carbonin commented on spec/lib/task_helpers/exports/alert_sets_spec.rb:
```diff
@@ -0,0 +1,21 @@
+  let(:alert) { 'Alert_Profile_VM_Import_Test.yaml' }
+
+  before(:each) do
+    @export_dir = Dir.mktmpdir('miq_exp_dir')
```
I think this should be a `let`

carbonin commented on spec/lib/task_helpers/exports/alert_sets_spec.rb:
```diff
@@ -0,0 +1,21 @@
+
+    options = { :directory => @export_dir }
+    TaskHelpers::Exports::AlertSets.new.export(options)
+    expect((Dir.entries(@export_dir) - %w(. ..))).not_to be_empty
```
I would rather assert on the contents of the export file rather than just if it exists.

carbonin commented on spec/lib/task_helpers/exports/alerts_spec.rb:
```diff
@@ -0,0 +1,21 @@
+
+    options = { :directory => @export_dir }
+    TaskHelpers::Exports::Alerts.new.export(options)
+    expect((Dir.entries(@export_dir) - %w(. ..))).not_to be_empty
```
Same here, assert on the contents rather than the file existence.

carbonin commented on spec/lib/task_helpers/exports_spec.rb:
```diff
@@ -0,0 +1,73 @@
+      expect(TaskHelpers::Exports.validate_directory(@export_dir2)).to eq('Destination directory must exist')
+    end
+
+    # rubocop:disable Style/NumericLiteralPrefix
```
Can we remove these?

carbonin commented on spec/lib/task_helpers/imports/alerts_spec.rb:
```diff
@@ -0,0 +1,26 @@
+describe 'TaskHelpers::Imports::Alerts' do
+  let(:data_dir) { File.join(File.expand_path(__dir__), 'data', 'alerts') }
+  let(:alert) { 'Alert_Import_Test.yaml' }
+  let(:bad_alert) { 'Bad_Alert_Import_Test.yml' }
```
Can we align these braces? I know it's not a rubocop thing, but it's something we usually do.

branic commented on spec/lib/task_helpers/exports/alert_sets_spec.rb:
```diff
@@ -0,0 +1,21 @@
+describe 'TaskHelpers::Exports::AlertSets' do
```
Fixed this. I was inconsistent with this, some strings some not.

branic commented on spec/lib/task_helpers/exports_spec.rb:
```diff
@@ -0,0 +1,73 @@
+describe TaskHelpers::Exports do
+  describe '.safe_filename' do
+    context 'with filename' do
```
Removed

branic commented on spec/lib/task_helpers/imports/alert_sets_spec.rb:
```diff
@@ -0,0 +1,54 @@
+describe 'TaskHelpers::Imports::AlertSets' do
```
Ugh, I thought i got all of those. It's fixed now.

gmcculloug commented:
@branic Can you modify the description of this PR to match the object type that is supported with the final version of the PR.

I would also suggest creating a git issue with a check list to track all the of the types so we can link to the PRs and check them off as we go.

branic commented:
@gmcculloug I've updated the description as well as opened an issue to track all of the object types.
